crates/mpz-common/src/executor/mt.rs

```diff
@@ -0,0 +1,213 @@
+use async_trait::async_trait;
+use scoped_futures::ScopedBoxFuture;
+use serio::IoDuplex;
+use uid_mux::FramedUidMux;
+
+use crate::{
+    context::{ContextError, ErrorKind},
+    Context, ThreadId,
+};
+
+/// A multi-threaded executor.
+#[derive(Debug)]
+pub struct MTExecutor<M> {
+    id: ThreadId,
+    mux: M,
+}
+
+impl<M> MTExecutor<M>
+where
+    M: FramedUidMux<ThreadId> + Clone,
+    M::Error: Into<Box<dyn std::error::Error + Send + Sync + 'static>>,
+{
+    /// Creates a new multi-threaded executor.
+    pub fn new(mux: M) -> Self {
+        Self {
+            id: ThreadId::default(),
+            mux,
+        }
+    }
+
+    /// Creates a new thread.
+    pub async fn new_thread(
+        &mut self,
+    ) -> Result<MTContext<M, <M as FramedUidMux<ThreadId>>::Framed>, ContextError> {
+        let id = self.id.increment().ok_or_else(|| {
+            ContextError::new(
+                ErrorKind::Thread,
+                "exceeded maximum number of threads (255)",
+            )
+        })?;
+
+        let io = self
+            .mux
+            .open_framed(&id)
+            .await
+            .map_err(|e| ContextError::new(ErrorKind::Mux, e))?;
+
+        Ok(MTContext {
+            id,
+            mux: self.mux.clone(),
+            io,
+            child: None,
+        })
+    }
+}
+
+/// A thread context from a multi-threaded executor.
+#[derive(Debug)]
+pub struct MTContext<M, Io> {
+    id: ThreadId,
+    mux: M,
+    io: Io,
+    // TODO: Support multiple children. Right now this is simpler to implement,
+    // and our `Context` trait only exposes joining two futures.
+    child: Option<Box<Self>>,
+}
+
+impl<M, Io> MTContext<M, Io> {
+    fn set_child(&mut self, child: Box<Self>) {
+        self.child = Some(child);
+    }
+}
+
+impl<M, Io> MTContext<M, Io>
+where
+    M: FramedUidMux<ThreadId, Framed = Io> + Clone,
+    M::Error: Into<Box<dyn std::error::Error + Send + Sync + 'static>>,
+{
+    async fn fork(&mut self) -> Result<Box<Self>, ContextError> {
+        // Forking a thread context is only performed once, afterwhich the child ctx
+        // is stored for later use.
+
+        if let Some(child) = self.child.take() {
+            return Ok(child);
+        }
```

Comment:
Shouldn't this be an error, because the caller does not know if he created a new thread or gets an existing one.

EDIT: I realize this method is not public, so I think it should be ok.